### ISSUE TITLE
Any load options break `with_entities()` in 1.4

### ISSUE THREAD

**Issue by ThiefMaster** [sqlalchemy/sqlalchemy#6253]
**Describe the bug**
Very similar to #6052 (which is supposedly fixed?), applying any loader option breaks `with_entities()`.

**Expected behavior**
No error, this output from the test script:

```
normal
[<Thing 1>]
with entities + noload
[(1,)]
with entities + joinedload
[(1,)]
```

**To Reproduce**

```py
from sqlalchemy import *
from sqlalchemy.ext.declarative import declarative_base  # to test on 1.3
from sqlalchemy.orm import *


Base = declarative_base()


class Thing(Base):
    __tablename__ = 'things'
    id = Column(Integer, primary_key=True)
    owner_id = Column(Integer, ForeignKey('users.id'), nullable=True, index=True)
    owner = relationship('User', backref='things')

    def __repr__(self):
        return f'<Thing {self.id}>'


class User(Base):
    __tablename__ = 'users'
    id = Column(Integer, primary_key=True)

    def __repr__(self):
        return f'<User {self.id}>'


e = create_engine('postgresql:///test', echo=False)
Base.metadata.create_all(e)
s = Session(e)

if not s.query(Thing).count():
    thing = Thing(owner=User())
    s.add(thing)
    s.commit()


print('normal')
print(s.query(Thing).all())
print('with entities + noload')  # fails
print(s.query(Thing).options(noload('owner')).with_entities(Thing.id).all())
print('with entities + joinedload')  # fails
print(s.query(Thing).options(joinedload('owner')).with_entities(Thing.id).all())
```

**Error**

```pytb
normal
[<Thing 1>]
with entities + noload
Traceback (most recent call last):
  File "/home/adrian/dev/indico/py3/src/satest.py", line 40, in <module>
    print(s.query(Thing).options(noload('owner')).with_entities(Thing.id).all())
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2685, in all
    return self._iter().all()
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/query.py", line 2820, in _iter
    result = self.session.execute(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 1670, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 313, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1381, in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 535, in _compile_w_cache
    compiled_sql = self._compiler(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 570, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 766, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 455, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 490, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/visitors.py", line 81, in _compiler_dispatch
    return meth(self, **kw)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/compiler.py", line 2966, in visit_select
    compile_state = select_stmt._compile_state_factory(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/sql/base.py", line 492, in create_for_statement
    return klass.create_for_statement(statement, compiler, **kw)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/context.py", line 579, in create_for_statement
    opt.process_compile_state(self)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/strategy_options.py", line 182, in process_compile_state
    self._process(compile_state, not bool(compile_state.current_path))
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/strategy_options.py", line 705, in _process
    val._bind_loader(
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/strategy_options.py", line 852, in _bind_loader
    entity = self._find_entity_basestring(entities, token, raiseerr)
  File "/home/adrian/dev/indico/py3/env/lib/python3.9/site-packages/sqlalchemy/orm/strategy_options.py", line 983, in _find_entity_basestring
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Query has only expression-based entities - can't find property named "owner".
```

**Additional context**

The example script is minimal, hence the "strange" usage of `with_entities`. In my real application the query is coming from a util function which applies a joinedload option, and overriding it with a noload doesn't help because *any* load option triggers this bug.

**Versions.**
 - OS: Linux
 - Python: 3.9.1
 - SQLAlchemy: 1.4.7
 - Database: Postgres
 - DBAPI: psycopg2

**Comment by zzzeek** [comment]
this is not a new behavior.  try this in 1.3, same error:

    s.query(Thing.id).options(noload("owner")).all()

the difference is that Query now processes all the options and everything at the end:  https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#many-core-and-orm-statement-objects-now-perform-much-of-their-construction-and-validation-in-the-compile-phase

this behavior is critical for performance.

I don't have a way to "fix" this other than just to ignore all invalid loader options in all cases, which seems bad also.

**Comment by zzzeek** [comment]
the best we could do is add some kind of "don't want for any options" flag if with_entities() is called, but that seems pretty hacky too.

**Comment by ThiefMaster** [comment]
An option/method to explicitly discard existing load options could be a less hacky option.... Anyway, in this particular case I managed to [just remove the option](https://github.com/indico/indico/commit/2b5d695e2c6fdc22d36b6b82c62e11c9e418c978) since it apparently wasn't needed anyway, but who knows what other cases may have the same problem...

**Comment by zzzeek** [comment]
well the bigger problem is the idea of "with_entities()" in the first place.   I want to make things simpler so that you just build a single query up front that's what you want rather than undoing state that was already created.

**Comment by ThiefMaster** [comment]
IMHO the nice thing of `with_entities()` is that you can have a util function that returns a query, and then use all the things it does, but query just some specific data in the end.

In my particular example where I encountered the bug: usually the function searches rooms matching various criteria and returns those rooms, but in some cases all I need is the ID and another column with generated data, so i'm using `with_entities()` to avoid getting all the other things I don't need.

**Comment by CaselIT** [comment]
IIRC you can use `load_only` for that.

At least I use `load_only` to specify which column to load, I think sqlalchemy will still always add the primary keys with `load_only`, not sure if with_entities is the same

**Comment by zzzeek** [comment]
well that gives you entities though, and those eagerload options if any will take effect.   it's more lightweight to load the column alone.

**Comment by CaselIT** [comment]
Oh I missed the part where only the column were loaded.

The documentation is not really clear about it.
To load tuples I usually just use a normal select as I would with a table even with mapped models. I guess in the end they would build the same thing

**Comment by zzzeek** [comment]
I can think of ways to make this work but they are all a little on the elaborate side.   The one that is least elaborate is to also store a "generation counter" on the query.    when the entities are changed (which is now the _raw_columns collection), we bump that number.   the loader options themselves would also need to be associated with a specific generation counter.   we detect the use case here by observing a mismatch in the generation and dont error in that case, but we still attempt to apply the options.

**Comment by CaselIT** [comment]
I noticed that the loaders use the strings, that are deprecated. Would using the relationship help in this case?

**Comment by zzzeek** [comment]
at the moment the loader strategies are pretty aggressive about pointing out things that dont make sense.  this was in response to angry users who werent using the options correctly and were completely confused, since the options gave no feedback at all if they made no sense.


    s.query(Thing).options(noload(User.id)).all()

    sqlalchemy.exc.ArgumentError: Mapped attribute "User.id" does not apply to any of the 
    root entities in this query, e.g. mapped class Thing->things. Please specify the full path 
    from one of the root entities to the target attribute.

that error message was not an accident :)   i was trying to spell things out as clearly as possible.

**Comment by zzzeek** [comment]
so maybe we keep messages that have to do with entity mismatch as above.    but then "only column entitiies", which is a separate message, we can soften on that, maybe if they were after the fact.

**Comment by zzzeek** [comment]
oh or this, with_entities has only column entities, then we set disable eagerloads.  that might do it.   in any case we need to add a litle bit of checking when the _raw_columns is changed.

**Comment by zzzeek** [comment]
but it has to apply to Select.with_only_columns also

**Comment by zzzeek** [comment]
so this would do that, however it means we wont catch an error for something like this:

    s.query(Thing.id).options(noload("owner")).with_entities(Thing.id).all()

I'm a little nervous of behaviors that flip on and off at a distance like that.

```diff
diff --git a/lib/sqlalchemy/orm/query.py b/lib/sqlalchemy/orm/query.py
index 9348d7d48f..15bbe1f3cb 100644
--- a/lib/sqlalchemy/orm/query.py
+++ b/lib/sqlalchemy/orm/query.py
@@ -125,6 +125,8 @@ class Query(
     _legacy_setup_joins = ()
     _label_style = LABEL_STYLE_LEGACY_ORM
 
+    _raw_columns_generation = 0
+
     _compile_options = ORMCompileState.default_compile_options
 
     load_options = QueryContext.default_load_options
@@ -1387,6 +1389,7 @@ class Query(
             return iter(())
         q = self._clone().enable_eagerloads(False)
         q._set_entities(columns)
+        q._raw_columns_generation += 1
         if not q.load_options._yield_per:
             q.load_options += {"_yield_per": 10}
         return iter(q)
@@ -1436,6 +1439,7 @@ class Query(
 
         """
         self._set_entities(entities)
+        self._raw_columns_generation += 1
 
     @_generative
     def add_columns(self, *column):
diff --git a/lib/sqlalchemy/orm/strategy_options.py b/lib/sqlalchemy/orm/strategy_options.py
index 4827b37524..b51aed1cd9 100644
--- a/lib/sqlalchemy/orm/strategy_options.py
+++ b/lib/sqlalchemy/orm/strategy_options.py
@@ -698,6 +698,9 @@ class _UnboundLoad(Load):
         self.__dict__ = state
 
     def _process(self, compile_state, raiseerr):
+        if compile_state.select_statement._raw_columns_generation > 0:
+            raiseerr = False
+
         dedupes = compile_state.attributes["_unbound_load_dedupes"]
         for val in self._to_bind:
             if val not in dedupes:
diff --git a/lib/sqlalchemy/sql/selectable.py b/lib/sqlalchemy/sql/selectable.py
index 2a221cdf0b..aeec4edf31 100644
--- a/lib/sqlalchemy/sql/selectable.py
+++ b/lib/sqlalchemy/sql/selectable.py
@@ -4476,6 +4476,7 @@ class Select(
     _having_criteria = ()
     _from_obj = ()
     _auto_correlate = True
+    _raw_columns_generation = 0
 
     _compile_options = SelectState.default_select_compile_options
 
@@ -4499,6 +4500,7 @@ class Select(
             ("_distinct", InternalTraversal.dp_boolean),
             ("_distinct_on", InternalTraversal.dp_clauseelement_tuple),
             ("_label_style", InternalTraversal.dp_plain_obj),
+            ("_raw_columns_generation", InternalTraversal.dp_plain_obj),
         ]
         + HasPrefixes._has_prefixes_traverse_internals
         + HasSuffixes._has_suffixes_traverse_internals
@@ -5376,6 +5378,7 @@ class Select(
                 c = c.self_group(against=operators.comma_op)
             rc.append(c)
         self._raw_columns = rc
+        self._raw_columns_generation += 1
 
     @property
     def whereclause(self):
```

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**memoize current options and joins w with_entities/with_only_cols** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2871

**Comment by zzzeek** [comment]
as noted in the above gerrit and issue #6503, I took the elaborate approach which is that the "_options" and the "_joins" etc. are all stuffed into a brand new kind of object that is pushed onto a stack whenever with_entities() or with_only_columns() is called.   that new object then participates in all the caching, adaption, etc. and serves as the previous version of the query before the entities were changed.   then the joins/options are processed in terms of the entities they were made against.
